contracts/utils/UniversalResolver.sol

```diff
@@ -7,51 +7,197 @@ import "./LowLevelCallUtils.sol";
+    function resolve(bytes calldata name, bytes memory data)
+        external
+        view
+        override
+        returns (bytes memory)
+    {
+        (Resolver resolver, ) = findResolver(name);
+        if (address(resolver) == address(0)) {
             return "";
         }
 
-        if(resolver.supportsInterface(type(IExtendedResolver).interfaceId)) {
-            return callWithOffchainLookupPropagation(
-                address(resolver),
-                abi.encodeWithSelector(IExtendedResolver.resolve.selector, name, data),
-                UniversalResolver.resolveCallback.selector
-            );
+        if (resolver.supportsInterface(type(IExtendedResolver).interfaceId)) {
+            return
+                callWithOffchainLookupPropagation(
+                    address(resolver),
+                    abi.encodeWithSelector(
+                        IExtendedResolver.resolve.selector,
+                        name,
+                        data
+                    ),
+                    UniversalResolver.resolveCallback.selector
+                );
         } else {
-            return callWithOffchainLookupPropagation(
-                address(resolver),
-                data,
-                UniversalResolver.resolveCallback.selector
-            );
+            return
+                callWithOffchainLookupPropagation(
+                    address(resolver),
+                    data,
+                    UniversalResolver.resolveCallback.selector
+                );
         }
     }
 
-    function supportsInterface(bytes4 interfaceId) public view virtual override returns(bool) {
-        return interfaceId == type(IExtendedResolver).interfaceId || super.supportsInterface(interfaceId);
+    struct ReverseData {
+        string dataType;
+        bytes data;
+    }
+
+    struct ReverseCall {
+        string sig;
+        ReverseData[] data;
+    }
+
+    /**
+     * @dev Performs ENS name reverse resolution for the supplied address and resolution data.
+     * @param name The address to resolve, in normalised and DNS-encoded form.
+     * @param calls The resolution data, as specified in ENSIP-10.
+     * @return The resolved name, and the resolved data.
+     */
+    function reverse(bytes calldata name, ReverseCall[] memory calls)
+        external
+        view
+        returns (string memory, bytes[] memory)
+    {
+        (Resolver resolver, bytes32 reverseNamehash) = findResolver(name);
+        if (address(resolver) == address(0)) {
+            return ("", new bytes[](0));
+        }
+
+        string memory resolvedName = resolver.name(reverseNamehash);
+        if (bytes(resolvedName).length == 0) {
+            return ("", new bytes[](0));
+        }
+
+        (bytes memory encodedName, bytes32 namehash) = resolvedName
+            .encodeAndHash();
+        bytes[] memory inputData = new bytes[](calls.length);
+
+        for (uint256 i = 0; i < calls.length; i++) {
```

Comment:
Wouldn't it be more effective to have the caller supply the already-ABI-encoded calldata, and just concatenate it?

Reply:
More generally, this batched approach is not going to play well with EIP 3668. It would probably make more sense to write a version of the function that does a single resolution operation, and we can wrap it in a batch function later.